plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -0,0 +1,570 @@
+            return true;
+        }
+
+        if (!StringUtils.isEmpty(pluginVersion)) {
+            pluginVersion = "latest";
+        }
+
+        if (!StringUtils.isEmpty(pluginUrl)) {
+            System.out.println("Will use url: " + pluginUrl);
+        } else if (pluginVersion.equals("latest") && !StringUtils.isEmpty(JENKINS_UC_LATEST)) {
+            urlString = new StringBuffer(JENKINS_UC_LATEST).append("/latest/").append(pluginName).append(".hpi").toString();
+        } else if (pluginVersion.equals("experimental")) {
+            urlString = new StringBuffer(JENKINS_UC_LATEST).append("/latest/").append(pluginName).append(".hpi").toString();
+        } else if (pluginVersion.contains("incrementals")) {
+            String[] incrementalsVersionInfo = pluginVersion.split(";");
+            String groupId = incrementalsVersionInfo[1];
+            String incrementalsVersion = incrementalsVersionInfo[2];
+
+            groupId = groupId.replace(".", "/");
+
+            String incrementalsVersionPath = new StringBuffer(incrementalsVersion).append("/").append(pluginName)
+                    .append("-").append(incrementalsVersion).append(".hpi").toString();
+
+            urlString = new StringBuffer(JENKINS_INCREMENTALS_REPO_MIRROR).append("/").append(groupId).append("/").
+                    append(incrementalsVersionPath).toString();
+        } else {
+            String pathToPlugin = new StringBuffer(pluginName).append("/").append(pluginVersion).append("/").
+                    append(pluginName).append(".hpi").toString();
+            urlString = new StringBuffer(JENKINS_UC_DOWNLOAD).append("/plugins/").append(pathToPlugin).toString();
+        }
+
+        System.out.println("Downloading plugin: " + pluginName + " from url: " + urlString);
+
+        File pluginFile = new File(refDir + SEPARATOR + plugin.getArchiveFileName());
+
+        try {
+            downloadToFile(urlString, pluginFile);
+        } catch (IOException | URISyntaxException e) {
+            e.printStackTrace();
+            failedPlugins.add(plugin);
+            System.out.println("Failed to download from requested URL");
+            return false;
+        }
+
+        //check integrity of plugin file
+        try {
+            JarFile pluginJpi = new JarFile(pluginFile);
+        } catch (IOException e) {
+            e.printStackTrace();
+            failedPlugins.add(plugin);
+            System.out.println("Downloaded file is not a valid ZIP");
+            return false;
+        }
+        return true;
+    }
+
+
+    public void downloadToFile(String urlString, File pluginFile) throws IOException, URISyntaxException {
+        CloseableHttpClient httpclient = HttpClients.createDefault();
+        HttpClientContext context = HttpClientContext.create();
+        HttpGet httpget = new HttpGet(urlString);
+        CloseableHttpResponse response = httpclient.execute(httpget, context);
+
+        try {
```

Comment:
you should be able to use try-with-resources to auto close the response, this means you won't need the finally block in this instance, but I think you need to close the http client as well, so you might need to keep it but change it to `httpclient.close()`
```suggestion
        try (CloseableHttpResponse response = httpclient.execute(httpget, context);) {
```

Reply:
Resolved in 7ca3d5d987f6c60617f5f95994574849fb17f170.